File: infer/tests/codetoanalyze/cpp/shared/methods/return_struct.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

struct X {
  int f;
};

struct A {
  X get(int p) {
    X x;
    return x;
  }
};

int test(A* a) {
  X x = a->get(1);
  return 1 / x.f;
}

File: infer/tests/codetoanalyze/objc/bufferoverrun/return_struct.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#import <Foundation/Foundation.h>

struct s {
  int a;
};

@interface ReturnStruct : NSObject
@end

@implementation ReturnStruct

- (struct s)get_struct:(int)n {
  struct s x;
  x.a = n;
  return x;
}

- (void)call_get_struct_Ok {
  int arr[5];
  struct s x = [self get_struct:3];
  arr[x.a] = 0;
}

- (void)call_get_struct_Bad {
  int arr[3];
  struct s x = [self get_struct:5];
  arr[x.a] = 0;
}

@end
